rpc/server.go

```diff
@@ -90,13 +90,16 @@ func (hdr *Header) IsRequest() bool {
 	return hdr.Request.Type != "" || hdr.Request.Action != ""
 }
 
-// ObserverFactory is a type which can construct a new Observer.
-type ObserverFactory interface {
-
-	// RPCObserver will return a new Observer usually constructed
-	// from the state previously built up in the Observer. The
-	// returned instance will be utilized per RPC request.
-	RPCObserver() Observer
+// RecorderFactory is a function that returns a recorder to record
+// details of a single request/response.
+type RecorderFactory func() Recorder
+
+// Recorder represents something the connection use to record requests
+// and replies. Recording a message can fail (for example for audit
+// logging), and when it does the request should be failed as well.
+type Recorder interface {
+	ServerRequest(hdr *Header, body interface{}) error
```

Comment:
Would prefer these method names start with a verb, eg
HandleServerRequest()
HandleServerReply()

or just

HandleRequest()
HandleReply()

the later nicely matches the name of the funcs in which these would be called

Reply:
I picked these to match the Observer interface methods (the only difference is the error return). But I think you're right it would be better for them to be verbs, and they don't actually need to match the Observer methods - maybe having them be distinct might make it clearer anyway. I'll change them.